### PR TITLE
CologneChip LUT tree mapping

### PR DESCRIPTION
**Pull request by gatecat**

### CODE REVIEW

File: techlibs/gatemate/synth_gatemate.cc [hunk 1]
```diff
@@ -303,11 +311,22 @@ struct SynthGateMatePass : public ScriptPass
...
+				run("abc " + abc_args, "(with -luttree)");
+				run("techmap -map +/gatemate/lut_tree_map.v", "(with -luttree)");
+				run("gatemate_foldinv", "(with -luttree)");
+				run("techmap -map +/gatemate/inv_map.v", "(with -luttree)");
+			} else {
+				std::string abc_args = " -dress -lut 4";
+				if (dff) {
+					abc_args += " -dff";
+				}
+				run("abc " + abc_args, "(without -luttree)");
```
Comment (Ravenslofty) [inline techlibs/gatemate/synth_gatemate.cc:328]: The comment here never actually gets printed, since the if branch triggers in help mode, right?

Comment (gatecat) [reply techlibs/gatemate/synth_gatemate.cc:329]: Good spot; should be fixed now to print both branches in help mode

**Comment by gatecat** [comment]
I have added random-generated tests for this in the most recent commit. This currently includes the generated files checked-in for simplicity but I can change it to generate them as part of running the tests if preferred.

**Comment by pu-cc** [comment]
The docs department has just defined consistent in- and output names for `CC_L2T4` and `CC_L2T5`. According to the most recent [Primtives Library Documentation](https://www.colognechip.com/docs/ug1001-gatemate1-primitives-library-2022-01.pdf), the final in- and output names are:

```
CC_L2T4      CC_L2T5
=========    =========

D0A -> I0    D0B -> I0
D1A -> I1    D1B -> I1
D2A -> I2    D2B -> I2
D3A -> I3    D3B -> I3
O   -> Y     D0  -> I4
             O   -> Y
```

**Comment by pu-cc** [comment]
The generated `lut_tree_map.v` contains:

```
[...]

module \$__CC4_AA_A (input A, B, C, E, output Y);
    CC_L2T5 #(
        .INIT_L02(4'b1000),
        .INIT_L03(4'b1010),
        .INIT_L11(4'b1000),
        .INIT_L20(4'b1000),
    ) _TECHMAP_REPLACE_ (
         .D0(E), .D0B(A), .D1B(B), .D2B(C), .D3B(),
         .O(Y)
    );
endmodule

module \$__CC4_OO_O (input A, B, C, E, output Y);
    CC_L2T5 #(
        .INIT_L02(4'b1110),
        .INIT_L03(4'b1010),
        .INIT_L11(4'b1110),
        .INIT_L20(4'b1110),
    ) _TECHMAP_REPLACE_ (
         .D0(E), .D0B(A), .D1B(B), .D2B(C), .D3B(),
         .O(Y)
    );
endmodule

module \$__CC4_XX_X (input A, B, C, E, output Y);
    CC_L2T5 #(
        .INIT_L02(4'b0110),
        .INIT_L03(4'b1010),
        .INIT_L11(4'b0110),
        .INIT_L20(4'b0110),
    ) _TECHMAP_REPLACE_ (
         .D0(E), .D0B(A), .D1B(B), .D2B(C), .D3B(),
         .O(Y)
    );
endmodule

[...]
```

Those can be mapped into L2T4. More specifically, they can actually be discarded as they match `$__CC4_{OOO,XXX,AAA}`. Or am I missing something?

**Comment by pu-cc** [comment]
Again in `lut_tree_map.v`:

```
module \$__CC_MUX (input A, B, C, output Y);
    CC_L2T4 #(
        .INIT_L00(4'b0010),
        .INIT_L01(4'b1000),
        .INIT_L10(4'b1110),
    ) _TECHMAP_REPLACE_ (
         .D0A(A), .D1A(C), .D2A(B), .D3A(C),
         .O(Y)
    );
endmodule
```

Can you please map `$__CC_MUX` to `CC_MX2` instead of `CC_L2T4`?

**Comment by gatecat** [comment]
Thanks for your comments, I think all of them have been addressed.

(with regards to the second point, I think the higher cost associated with these variants should mean they're never actually mapped. but I agree it's cleaner not to have them there in the first place).

**Comment by pu-cc** [comment]
Looks good, thank you! I'll be working with the new netlists on our PnR for the next few days and will let you know if I notice anything else.

**Comment by pu-cc** [comment]
Apologies it had to wait. `CC_L2T4` and `CC_L2T5` primitive support is now available in P&R -- you are welcome merge if there are no other objections from your side. Thanks again!

**Comment by gatecat** [comment]
Thanks for confirming, this is merged now!

**Pull request merged**